`0`...`9` caused a C error

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V doctor:**
```
OS: windows, Microsoft Windows 10 Home v19042 64-bit
Processor: 2 cpus, 64bit, little endian, AMD A6-9225 RADEON R4, 5 COMPUTE CORES 2C+3G
CC version: Error: exec failed (CreateProcess) with code 2: The system cannot find the file specified.
 cmd: cc --version

getwd: C:\Users\muham\Documents\vscode workspace\dewowo\calc
vmodules: C:\Users\muham\.vmodules
vroot: C:\v
vexe: C:\v\v.exe
vexe mtime: 2021-08-30 10:06:54
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 f8d6fe7.f44eb88

Git version: git version 2.31.1.windows.1
Git vroot status: weekly.2021.33.2-104-gf44eb88a
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 0f5168cd
```

**What did you do?**
In the match block
```v
`0`...`9` {
    result << Token{
	t_type: Toktype.integer
	i_value: char.int()
    }
}
```
the generated C file
```c
...
else if (string__eq(_v_char, _SLIT("/"))) {
	array_push((array*)&result, _MOV((main__T...
	continue;
}
else if (string__eq(_v_char, )) {
                            ^-- this one
	array_push((array*)&result, _MOV((main__T...
}
```

**What did you expect to see?**
No error
 
**What did you see instead?**
==================
```
C:/Users/muham/AppData/Local/Temp/v/calc.17808378873513117939.tmp.c:14201: warning: assignment from incompatible pointer type
C:/Users/muham/AppData/Local/Temp/v/calc.17808378873513117939.tmp.c:14412: error: identifier expected
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -6223,14 +6223,14 @@ fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSym
 				low_expr := expr.low
 				high_expr := expr.high
 				if low_expr is ast.IntegerLiteral {
-					if high_expr is ast.IntegerLiteral {
+					if high_expr is ast.IntegerLiteral && cond_type_sym.is_int() {
 						low = low_expr.val.i64()
 						high = high_expr.val.i64()
 					} else {
 						c.error('mismatched range types', low_expr.pos)
 					}
 				} else if low_expr is ast.CharLiteral {
-					if high_expr is ast.CharLiteral {
+					if high_expr is ast.CharLiteral && cond_type_sym.kind in [.byte, .char, .rune] {
 						low = low_expr.val[0]
 						high = high_expr.val[0]
 					} else {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -6223,7 +6223,8 @@ fn (mut c Checker) match_exprs(mut node ast.MatchExpr, cond_type_sym ast.TypeSym
 				low_expr := expr.low
 				high_expr := expr.high
 				if low_expr is ast.IntegerLiteral {
-					if high_expr is ast.IntegerLiteral && cond_type_sym.is_int() {
+					if high_expr is ast.IntegerLiteral
+						&& (cond_type_sym.is_int() || cond_type_sym.info is ast.Enum) {
 						low = low_expr.val.i64()
 						high = high_expr.val.i64()
 					} else {

--- a/vlib/v/checker/tests/match_range_mismatch_type_err.out
+++ b/vlib/v/checker/tests/match_range_mismatch_type_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/match_range_mismatch_type_err.vv:4:3: error: mismatched range types
+    2 |     x := '1'
+    3 |     match x {
+    4 |         `0`...`9` {
+      |         ~~~
+    5 |             println('0-9')
+    6 |         }

--- a/vlib/v/checker/tests/match_range_mismatch_type_err.vv
+++ b/vlib/v/checker/tests/match_range_mismatch_type_err.vv
@@ -1,0 +1,11 @@
+fn main() {
+	x := '1'
+	match x {
+		`0`...`9` {
+			println('0-9')
+		}
+		else {
+			println('!0-9')
+		}
+	}
+}

